### PR TITLE
Add Bayesian Personalized Ranking Loss for link prediction

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*
Add Bayesian Personalized Ranking Loss (https://arxiv.org/abs/1205.2618) for link prediction

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/test_gsf.py
Comment (thvasilo) [inline tests/unit-tests/test_gsf.py:1]: Any chance we can add some tests for the logic of the loss? Something like pre-computed values from another source (e.g. simple scipy/numpy implementation)

Comment (classicsong) [reply tests/unit-tests/test_gsf.py:1]: added

File: docs/source/advanced/link-prediction.rst [hunk 1]
```diff
@@ -159,6 +158,26 @@ GraphStorm provides four options to compute training losses:
...
+
+        x = positive\_score - negative\_score
+
+* **Weighted Bayesian Personalized Ranking Loss**: The weighted bayesian personalized ranking is similar to **Bayesian Personalized Ranking Loss** except that it allows users to set a weight for each positive edge. The loss is defined as:
+
+    .. math::
+
+        loss = - w\_e [ \log\left(\frac{ 1 }{ 1 + \exp(-x)}\right) ]
+
+    where ``x`` is the distance between the score of a positive edge and the score of its corresponding negative edge. `w_e`` is the weight of the positive edge.
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:179]:
```suggestion
    where ``x`` is the distance between the score of a positive edge and the score of its corresponding negative edge, and `w_e`` is the weight of the positive edge.
```

File: python/graphstorm/gsf.py [hunk 2]
```diff
@@ -781,6 +784,9 @@ def create_builtin_lp_gnn_model(g, config, train_task):
 def create_builtin_lp_decoder(g, decoder_input_dim, config, train_task):
     """ create builtin link prediction decoder according to task config
 
+    .. versionchanged:: 0.4.1
+        Add bayesian personalized ranking loss support
```
Comment (Oxfordblue7) [inline python/graphstorm/gsf.py:788]:
```suggestion
        Add bayesian personalized ranking loss support.
```

File: python/graphstorm/model/loss_func.py [hunk 3]
```diff
@@ -550,6 +552,163 @@ def forward(self, pos_score, neg_score):
...
+    r""" Loss function for link prediction tasks using bayesian
+    personalized ranking (BPR) (https://arxiv.org/abs/1205.2618) loss.
+
+    The loss function is defined as:
+
+    .. math::
+
+        loss = - \log\left(\frac{ 1 }{ 1 + \exp(-x)}\right)
+
+    where ``x`` is positive_score - negative_score
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:565]:
```suggestion
    where ``x`` is positive_score - negative_score.
```

File: python/graphstorm/model/loss_func.py [hunk 4]
```diff
@@ -550,6 +552,163 @@ def forward(self, pos_score, neg_score):
...
+
+    @property
+    def in_dims(self):
+        """ The number of input dimensions.
+
+        Returns
+        -------
+        int : the number of input dimensions.
+        """
+        return None
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:612]: Why "None"?

File: python/graphstorm/model/loss_func.py [hunk 5]
```diff
@@ -550,6 +552,163 @@ def forward(self, pos_score, neg_score):
...
+
+    @property
+    def out_dims(self):
+        """ The number of output dimensions.
+
+        Returns
+        -------
+        int : the number of output dimensions.
+        """
+        return None
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:622]: Why "None"?

File: python/graphstorm/model/loss_func.py [hunk 6]
```diff
@@ -550,6 +552,163 @@ def forward(self, pos_score, neg_score):
...
+    @property
+    def in_dims(self):
+        """ The number of input dimensions.
+
+        Returns
+        -------
+        int : the number of input dimensions.
+        """
+        return None
+
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:699]: ditto

File: python/graphstorm/model/loss_func.py [hunk 7]
```diff
@@ -550,6 +552,163 @@ def forward(self, pos_score, neg_score):
...
+    @property
+    def out_dims(self):
+        """ The number of output dimensions.
+
+        Returns
+        -------
+        int : the number of output dimensions.
+        """
+        return None
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:708]: ditto

File: tests/unit-tests/test_gsf.py [hunk 8]
```diff
@@ -43,7 +45,9 @@
                                            LinkPredictWeightedDistMultDecoder,
                                            LinkPredictRotatEDecoder,
                                            LinkPredictContrastiveRotatEDecoder,
-                                           LinkPredictWeightedRotatEDecoder)
+                                           LinkPredictWeightedRotatEDecoder,LinkPredictTransEDecoder,
```
Comment (Oxfordblue7) [inline tests/unit-tests/test_gsf.py:48]:
```suggestion
                                           LinkPredictWeightedRotatEDecoder,
                                           LinkPredictTransEDecoder,
```

File: python/graphstorm/model/loss_func.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:622]: In_dim and out_dim is required by `GSLayer`, but they are meaningless for lossFuncs. Just set them to None.

**Review by Oxfordblue7 — APPROVED** [review]
